File: Civitas - Ruby/lib/sorpresa_salir_carcel.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Sorpresa_salir_carcel < Sorpresa
    public_class_method :new
    def initialize mazo
      @mazo = mazo
    end
    
    #Aplica la carta para salir de la carcel
    def aplicar_a_jugador(actual, todos)
      salvocond = false
      if(jugador_correcto(actual, todos))
        informe(actual, todos)
        todos.each do |jugador_actual|
          if(jugador_actual.tiene_salvoconducto)
            salvocond = true
          end
        end
        if(salvocond)
          sorpresa = Sorpresa_salir_carcel.new(@mazo)
          todos[actual].obtener_salvoconducto(sorpresa)
          sorpresa.salir_del_mazo          
        end
      end
    end
    
    #Si el tipo de sorpresa es evitar la carcel, se inhabilita en el mazo de cartas sorpresa
    def salir_del_mazo
      @mazo.inhabilitar_carta_especial(self)
    end
    
    #Si el tipo de sorpresa es evitar la carcel, se habilita en el mazo de cartas sorpresa
    def usada
      @mazo.habilitar_carta_especial(self)
    end
    
  end
end

File: Civitas - Java/src/civitas/SorpresaSalirCarcel.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author juanc
 */
public class SorpresaSalirCarcel extends Sorpresa {
    protected MazoSorpresas mazo;
    
    //Para evitar ir a la carcel
    SorpresaSalirCarcel (MazoSorpresas mazo){
        super(-1,"SalirCarcel");
        this.mazo = mazo;
    }
    
    //Aplica la carta para salir de la carcel
    @Override
    protected void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        
        boolean salvocond = false;
        if(jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            
            for(int i = 0; i < todos.size() && !salvocond; i++)
                if(todos.get(i).tieneSalvoconducto())
                    salvocond = true;
            
            if(salvocond){
                SorpresaSalirCarcel sorpresa = new SorpresaSalirCarcel(mazo);
                todos.get(actual).obtenerSalvoconducto(sorpresa);
                sorpresa.salirDelMazo();
            }
            
        }
        
    }
    
    //Si el tipo de sorpresa es evitar la carcel, se inhabilita en el mazo de cartas sorpresa
    void salirDelMazo(){
        mazo.inhabilitarCartaEspecial(this);
        
    }
    
   //Si el tipo de sorpresa es evitar la carcel, se habilita en el mazo de cartas sorpresa
    void usada(){
        mazo.habilitarCartaEspecial(this);     
    }
    
}
